crates/store/README.md

```diff
@@ -0,0 +1,829 @@
+values of type `StateData`, providing compile-time guarantees about the types of
+data being processed.
+
+#### Integration with layers and transactions
+
+The iterator system is deeply integrated with other components of the library:
+
+  1. **Layer integration**: The `ReadLayer` trait includes an iter method that
+     returns an `Iter`. This allows each layer to provide its own implementation
+     of iteration, potentially adding additional functionality or optimisations.
+
+  2. **Transaction integration**: The `Transaction` struct provides methods for
+     creating iterators over its operations. This allows for efficient
+     processing of pending changes in a transaction.
+
+#### Range queries
+
+The iterator system supports efficient range queries through the use of a
+`start` key in the `iter()` method of `ReadLayer`. This allows for iteration to
+begin at a specific point in the key space, enabling efficient partial scans of
+the data.
+
+#### Performance considerations
+
+The iterator system is designed with performance in mind:
+
+  1. **Lazy evaluation**: Iterators are lazy, meaning they only do work when
+     asked for the next item. This allows for efficient processing of large
+     datasets without unnecessary memory usage.
+
+  2. **Zero-cost abstractions**: The use of Rust's trait system and generic
+     programming allows many of the abstractions to be resolved at compile-time,
+     resulting in efficient machine code.
+
+  3. **Custom implementations**: The `DBIter` trait allows for database-specific
+     optimisations in the underlying iteration logic.
+
+#### Extensibility
+
+The iterator system is designed to be extensible:
+
+  1. **Custom iterators**: New iterator types can be easily created by
+     implementing the `DBIter` trait.
+
+  2. **Composition**: The `IterPair` struct allows for the composition of
+     multiple iterators, enabling complex iteration patterns.
+
+#### Use cases
+
+The iterator system supports a wide range of use cases:
+
+  1. **Scanning**: Efficiently scan through large portions of the data store.
+
+  2. **Filtering**: Iterate over data and filter based on key or value
+     properties.
+
+  3. **Aggregation**: Perform aggregations or reductions over ranges of data.
+
+  4. **Prefix queries**: Efficiently query all keys with a certain prefix.
+
+#### Limitations and considerations
+
+While powerful, users of the iterator system should be aware of certain
+considerations:
+
+  1. **Consistency**: In a multi-threaded environment, iterators provide a
+     point-in-time view and may not reflect concurrent modifications. If an
+     iterator is to be shared across threads, care should be taken to ensure
+     consistency — and it may be preferable to use [Rayon][]'s parallel
+     iterators instead for concurrent processing.
+
+  2. **Resource usage**: Long-running iterations may hold resources for extended
+     periods. Care should be taken in concurrent scenarios.
```

Comment:
by resources you mean? potentially deleted values on the live view of the storage? or what?

Reply:
No... anything that's in use by the iterator. So, borrowed data, or anything in scope of the iterator logic. In a situation where that might be used across multiple threads, that can lead to thread starvation.

Reply:
Ah, so this is directed at people implementing `Database`?

Reply:
Well, this whole README is about the store. Regardless of who implements `Database`, iterating through data may cause concurrency issues if the things being held by the iteration process are required elsewhere. It is simply something to be aware of.